lib/evaluate/fold.cc

```diff
@@ -152,35 +157,486 @@ ComplexPart FoldOperation(FoldingContext &context, ComplexPart &&complexPart) {
+  }
+  return false;
+}
+
+template<typename TO, typename Operand>
+Scalar<TO> ScalarConvert(
+    FoldingContext &context, const Scalar<Operand> &value) {
+  static_assert(IsAllowedBasicConversion<TO::category, Operand::category>());
+  if constexpr (TO::category == TypeCategory::Integer) {
+    if constexpr (Operand::category == TypeCategory::Integer) {
+      auto converted{Scalar<TO>::ConvertSigned(value)};
+      if (converted.overflow) {
+        context.messages.Say(
+            "INTEGER(%d) to INTEGER(%d) conversion overflowed"_en_US,
+            Operand::kind, TO::kind);
+      }
+      return std::move(converted.value);
+    } else if constexpr (Operand::category == TypeCategory::Real) {
+      auto converted{value.template ToInteger<Scalar<TO>>()};
+      if (converted.flags.test(RealFlag::InvalidArgument)) {
+        context.messages.Say(
+            "REAL(%d) to INTEGER(%d) conversion: invalid argument"_en_US,
+            Operand::kind, TO::kind);
+      } else if (converted.flags.test(RealFlag::Overflow)) {
+        context.messages.Say(
+            "REAL(%d) to INTEGER(%d) conversion overflowed"_en_US,
+            Operand::kind, TO::kind);
+      }
+      return std::move(converted.value);
+    }
+  } else if constexpr (TO::category == TypeCategory::Real) {
+    char buffer[64];
+    if constexpr (Operand::category == TypeCategory::Integer) {
+      auto converted{Scalar<TO>::FromInteger(value)};
+      if (!converted.flags.empty()) {
+        std::snprintf(buffer, sizeof buffer,
+            "INTEGER(%d) to REAL(%d) conversion", Operand::kind, TO::kind);
+        RealFlagWarnings(context, converted.flags, buffer);
+      }
+      return std::move(converted.value);
+    } else if constexpr (Operand::category == TypeCategory::Real) {
+      auto converted{Scalar<TO>::Convert(value)};
+      if (!converted.flags.empty()) {
+        std::snprintf(buffer, sizeof buffer, "REAL(%d) to REAL(%d) conversion",
+            Operand::kind, TO::kind);
+        RealFlagWarnings(context, converted.flags, buffer);
+      }
+      if (context.flushSubnormalsToZero) {
+        converted.value = converted.value.FlushSubnormalToZero();
+      }
+      return std::move(converted.value);
+    }
+  } else if constexpr (TO::category == TypeCategory::Logical &&
+      Operand::category == TypeCategory::Logical) {
+    return value.IsTrue();
+  }
+}
+
+// helpers to fold intrinsic function references
+namespace intrinsicHelper {
+// helper to produce hash of intrinsic names based the first 3 letters. All
+// intrinsic names are longer than 3 letters
+static constexpr inline std::int32_t CommonHash(const char *s, std::size_t n) {
+  if (n < 3) {
+    return 0;
+  }
+  return (((static_cast<std::int32_t>(s[0]) << 8) + s[1]) << 8) + s[2];
+}
+
+static constexpr std::int32_t operator"" _hash(const char *s, std::size_t n) {
+  return CommonHash(s, n);
+}
+
+static std::int32_t DynamicHash(const std::string &s) {
```

Comment:
Hash is maybe not a really good name, the function could be called ThreeFirstLettersAsInt. It is just a way to help organizing the dispatching from the intrinsic name by switching on the first three letters (intrinsic are all at least three letters). I could also go letter by letter to avoid this kind of helper function but then I find the code harder to read because lines are too short due to indentation.
Finally, I could just sequentially go through the name, I went for the switch because I found it a bit more readable and in theory faster. So far, with this few functions, I cannot claim any real performance benefit. Do you think it is overkill and does not help reading?

Reply:
If you're going to hash, try hashing the whole name, use `constexpr` hashes as `case` labels, and confirming the name match in the case.  That seems readable and maintainable to me.

```
switch (IntrinsicNameHash(name, length)) {
case "acos"_intrinsicHash:
  if (IntrinsicNamesMatch("acos", name, length)) {
  }
  break;
...
}
```

The `IntrinsicNamesMatch()` predicate can be optimized under the assumption that the name and the string have the same hash.  If hashes are guaranteed to be perfect up to some length, then `IntrinsicNamesMatch()` can just return true when the length is short enough.  Example hash:

```
constexpr std::uint64_t IntrinsicNameHash(const char *name, std::size_t len) {
  CHECK(len <= 255);
  auto hash{static_cast<std::uint64_t>(len) << 56};
  for (; len > 0; --len) {
    hash ^= (static_cast<std::uint64_t>(*name++) & 255) << (8*(len % 7));
  }
  return hash;
}
```

Reply:
Thanks for the example, for now I have removed the hash and switch in order to focus the new PR #292 on the host runtime indirection feature. I will try using it if we go for the hash later.